File: pet_flutter/pet_web/src/services/reviewService.js
import axiosClient from '../utils/axiosClient';

class ReviewService {
  // Lấy tất cả reviews
  async getAllReviews() {
    try {
      const response = await axiosClient.get('/Reviews');
      return response.data;
    } catch (error) {
      console.error('Error fetching all reviews:', error);
      throw error;
    }
  }

  // Lấy review theo ID
  async getReviewById(reviewId) {
    try {
      const response = await axiosClient.get(`/Reviews/${reviewId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching review by ID:', error);
      throw error;
    }
  }

  // Lấy reviews của sản phẩm
  async getProductReviews(productId, sortBy = 'newest') {
    try {
      const response = await axiosClient.get(`/Reviews/Product/${productId}?sortBy=${sortBy}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching product reviews:', error);
      throw error;
    }
  }

  // Lấy reviews của sản phẩm với filter (alias cho getProductReviews)
  async getProductReviewsWithFilter(productId, sortBy = 'newest') {
    return this.getProductReviews(productId, sortBy);
  }

  // Lấy điểm đánh giá trung bình của sản phẩm
  async getProductRating(productId) {
    try {
      const response = await axiosClient.get(`/Reviews/Product/${productId}/Rating`);
      return response.data;
    } catch (error) {
      console.error('Error fetching product rating:', error);
      throw error;
    }
  }

  // Lấy reviews của user
  async getUserReviews(userId) {
    try {
      const response = await axiosClient.get(`/Reviews/User/${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching user reviews:', error);
      throw error;
    }
  }

  // Lấy reviews của đơn hàng
  async getOrderReviews(orderId) {
    try {
      const response = await axiosClient.get(`/Reviews/Order/${orderId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching order reviews:', error);
      throw error;
    }
  }

  // Tạo review mới
  async createReview(reviewData) {
    try {
      const response = await axiosClient.post('/Reviews', reviewData);
      return response.data;
    } catch (error) {
      console.error('Error creating review:', error);
      throw error;
    }
  }

  // Cập nhật review
  async updateReview(reviewId, reviewData) {
    try {
      const response = await axiosClient.put(`/Reviews/${reviewId}`, reviewData);
      return response.data;
    } catch (error) {
      console.error('Error updating review:', error);
      throw error;
    }
  }

  // Xóa review
  async deleteReview(reviewId) {
    try {
      await axiosClient.delete(`/Reviews/${reviewId}`);
      return true;
    } catch (error) {
      console.error('Error deleting review:', error);
      throw error;
    }
  }

  // Kiểm tra user đã đánh giá sản phẩm chưa
  async hasUserReviewedProduct(productId, userId) {
    try {
      const reviews = await this.getProductReviews(productId);
      return reviews.some(review => review.userId === userId);
    } catch (error) {
      console.error('Error checking user review status:', error);
      return false;
    }
  }

  // Kiểm tra user đã đánh giá đơn hàng chưa
  async hasUserReviewedOrder(orderId, userId) {
    try {
      const reviews = await this.getOrderReviews(orderId);
      return reviews.some(review => review.userId === userId);
    } catch (error) {
      console.error('Error checking user order review status:', error);
      return false;
    }
  }

  // Lấy review của user cho sản phẩm cụ thể
  async getUserProductReview(productId, userId) {
    try {
      const reviews = await this.getProductReviews(productId);
      return reviews.find(review => review.userId === userId) || null;
    } catch (error) {
      console.error('Error fetching user product review:', error);
      return null;
    }
  }

  // Lấy thống kê đánh giá của sản phẩm
  async getProductReviewStats(productId) {
    try {
      const reviews = await this.getProductReviews(productId);
      const rating = await this.getProductRating(productId);
      
      const stats = {
        totalReviews: reviews.length,
        averageRating: rating.averageRating || 0,
        ratingDistribution: {
          5: reviews.filter(r => r.rating === 5).length,
          4: reviews.filter(r => r.rating === 4).length,
          3: reviews.filter(r => r.rating === 3).length,
          2: reviews.filter(r => r.rating === 2).length,
          1: reviews.filter(r => r.rating === 1).length,
        }
      };
      
      return stats;
    } catch (error) {
      console.error('Error fetching product review stats:', error);
      throw error;
    }
  }

  // Lấy trạng thái mua hàng của user cho sản phẩm cụ thể (private - yêu cầu authorization)
  async getUserProductPurchaseStatus(userId, productId) {
    try {
      const response = await axiosClient.get(`/Reviews/User/${userId}/Product/${productId}/PurchaseStatus`);
      return response.data;
    } catch (error) {
      console.error('Error fetching user product purchase status:', error);
      throw error;
    }
  }

  // Lấy trạng thái mua hàng public (không yêu cầu authorization)
  async getPublicUserProductPurchaseStatus(userId, productId) {
    try {
      const response = await axiosClient.get(`/Reviews/Public/User/${userId}/Product/${productId}/PurchaseStatus`);
      return response.data;
    } catch (error) {
      console.error('Error fetching public user product purchase status:', error);
      throw error;
    }
  }

  // Kiểm tra user có thể đánh giá sản phẩm không (đã mua và đơn hàng hoàn tất)
  async canUserReviewProduct(userId, productId) {
    try {
      const status = await this.getPublicUserProductPurchaseStatus(userId, productId);
      return status.hasPurchased && (status.orderStatusDisplay === 'Hoàn thành' || status.orderStatusDisplay === 'Completed');
    } catch (error) {
      console.error('Error checking if user can review product:', error);
      return false;
    }
  }
}

const reviewService = new ReviewService();
export default reviewService; 

File: pet_flutter/lib/services/review_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import '../models/review.dart';
import 'secure_storage.dart';

class ReviewService {
  static String get _baseUrl => '${ApiConfig.baseUrl}/api/Reviews';

  // Lấy token từ secure storage
  Future<String?> _getAuthToken() async {
    final secureStorage = SecureStorageService();
    return await secureStorage.readToken();
  }

  // Tạo headers với authorization
  Future<Map<String, String>> _getHeaders() async {
    final token = await _getAuthToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  // Lấy tất cả reviews
  Future<List<Review>> getAllReviews() async {
    try {
      final response = await http.get(
        Uri.parse(_baseUrl),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((json) => Review.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load reviews: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting reviews: $e');
    }
  }

  // Lấy review theo ID
  Future<Review?> getReviewById(int reviewId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/$reviewId'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        return Review.fromJson(json.decode(response.body));
      } else if (response.statusCode == 404) {
        return null;
      } else {
        throw Exception('Failed to load review: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting review: $e');
    }
  }

  // Lấy reviews theo user ID
  Future<List<Review>> getReviewsByUserId(int userId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/User/$userId'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((json) => Review.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load user reviews: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting user reviews: $e');
    }
  }

  // Lấy reviews theo service ID
  Future<List<Review>> getReviewsByServiceId(int serviceId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/Service/$serviceId'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((json) => Review.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load service reviews: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting service reviews: $e');
    }
  }

  // Lấy reviews theo appointment ID
  Future<List<Review>> getReviewsByAppointmentId(int appointmentId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/Appointment/$appointmentId'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((json) => Review.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load appointment reviews: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting appointment reviews: $e');
    }
  }

  // Lấy reviews theo product ID
  Future<List<Review>> getReviewsByProductId(int productId, {String sortBy = 'newest'}) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/Product/$productId?sortBy=$sortBy'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((json) => Review.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load product reviews: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting product reviews: $e');
    }
  }

  // Lấy rating trung bình của service
  Future<double> getServiceRating(int serviceId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/Service/$serviceId/Rating'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> json = jsonDecode(response.body);
        return (json['averageRating'] ?? 0.0).toDouble();
      } else {
        throw Exception('Failed to load service rating: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting service rating: $e');
    }
  }

  // Lấy rating trung bình của product
  Future<double> getProductRating(int productId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/Product/$productId/Rating'),
        headers: await _getHeaders(),
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> json = jsonDecode(response.body);
        return (json['averageRating'] ?? 0.0).toDouble();
      } else {
        throw Exception('Failed to load product rating: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting product rating: $e');
    }
  }

  // Tạo review mới
  Future<Review> createReview(CreateReviewRequest request) async {
    try {
      
      final response = await http.post(
        Uri.parse(_baseUrl),
        headers: await _getHeaders(),
        body: json.encode(request.toJson()),
      );


      if (response.statusCode == 201) {
        if (response.body.isNotEmpty) {
          return Review.fromJson(json.decode(response.body));
        } else {
          throw Exception('Empty response body');
        }
      } else {
        String errorMessage = 'Failed to create review: ${response.statusCode}';
        
        if (response.body.isNotEmpty) {
          try {
            final errorBody = json.decode(response.body);
            errorMessage = errorBody['message'] ?? errorBody['title'] ?? errorMessage;
          } catch (e) {
            // If response body is not JSON, use the raw body
            errorMessage = 'Server error: ${response.body}';
          }
        }
        
        throw Exception(errorMessage);
      }
    } catch (e) {
      throw Exception('Error creating review: $e');
    }
  }

  // Cập nhật review
  Future<Review> updateReview(int reviewId, UpdateReviewRequest request) async {
    try {
      final response = await http.put(
        Uri.parse('$_baseUrl/$reviewId'),
        headers: await _getHeaders(),
        body: json.encode(request.toJson()),
      );

      if (response.statusCode == 200) {
        return Review.fromJson(json.decode(response.body));
      } else {
        final errorBody = json.decode(response.body);
        throw Exception('Failed to update review: ${errorBody['message'] ?? response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error updating review: $e');
    }
  }

  // Xóa review
  Future<bool> deleteReview(int reviewId) async {
    try {
      final response = await http.delete(
        Uri.parse('$_baseUrl/$reviewId'),
        headers: await _getHeaders(),
      );

      return response.statusCode == 204;
    } catch (e) {
      throw Exception('Error deleting review: $e');
    }
  }

  // Upload ảnh cho review
  Future<List<String>> uploadReviewImages(List<File> imageFiles) async {
    try {
      final token = await _getAuthToken();
      final request = http.MultipartRequest(
        'POST',
        Uri.parse('$_baseUrl/upload-images'),
      );

      if (token != null) {
        request.headers['Authorization'] = 'Bearer $token';
      }

      for (int i = 0; i < imageFiles.length; i++) {
        request.files.add(
          await http.MultipartFile.fromPath(
            'files',
            imageFiles[i].path,
          ),
        );
      }

      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((url) => url.toString()).toList();
      } else {
        throw Exception('Failed to upload images: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error uploading images: $e');
    }
  }
}
